ensure that we have an import_std configuration in the test-set

burnpanck opened the pull request:
also, update flake8

mpusz commented:
Ahh, I forgot to say that it also requires C++23 to work 😢 

Also, I do not know why a freestanding build failed.

FYI, I am providing a 3-day training from today, so I may be less responsive than usual.

mpusz commented:
I've just changed the CI scripts (but not the generator) to account for C++23. They should also check for something like `import_std_support` when provided by the matrix.

burnpanck commented:
the freestanding build failed because of an issue you seemed to have previously identified: In Debug builds, the CMakeCheckCompiler test fails due to a missing _main symbol. you had an "exclude" in the matrix, but "include" comes after, so we have to account for that in the generator. I forgot about that when I added that special "cxx_modules" subsample.

burnpanck commented:
I might try to merge freestanding with the hosted conan, making this just another dimension of the matrix. I'll also extract the cxx_modules into its own dimension, and ensure all the unsupported combinations are skipped.

burnpanck commented:
finally, I'll experiment with a truly random sample of configurations, together with a GitHub Actions "manual_trigger" with a user input to choose the random seed. I belive that should be possible.

mpusz commented:
WOW! You are a true GitHub Actions Wizard! 😄 I would never come up with those solutions by myself. It is great to learn those things from you. Thanks!

mpusz commented:
Also, it seems that there is some issue with `contracts` setting as it resolves to an empty string instead of "none" here:
https://github.com/mpusz/mp-units/actions/runs/11826393072/job/32952538814#step:15:3

mpusz commented:
Hi @burnpanck, will you have some time to look into it soon?

burnpanck commented:
I'l try to do it today. I would probably remove all logic affecting the settings of `std_format` and `import_std` from the CI, and instead include that into the generator; this has a few advantages: The logic determining which elements to run from the matrix properly respects what is actually being run in CI, and the logic determining what configurations are compatible is next to where the combinations are being configured. Would that be ok for you @mpusz? From my experience that would also be safer in the sense that a wrong configuration of the matrix would cause a loud failure rather than silently skipping an important test.

mpusz commented:
Sure, it sounds great. However, please be careful with the `import_std` as this requires a very specific configuration and it is enforced through the `conanfile.py`. If `import_std` is set to `True` but other options are invalid, Conan will raise an error.

burnpanck commented:
> If import_std is set to True but other options are invalid, Conan will raise an erro

That is a good thing. It means that we won't accidentally avoid testing `import_std=True`. If the matrix generator script thinks that a particular configuration with `import_std=True` is valid and schedules it for testing, it should either indeed test `import_std=True` or it should fail. Note also that the logic dealing with mutually incompatible configurations is very localised, so it should be fairly easy to maintain https://github.com/mpusz/mp-units/blob/bc440cc15f159b7ad8e017f8c0d9db6badb76fa7/.github/generate-job-matrix.py#L55-L72

Finally, as promised, I added changed the random seed to be truly random. It chosen seed is printed at the start of the "generate-matrix" job. I also added a [`workflow_dispatch`](https://docs.github.com/en/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows#workflow_dispatch) trigger to allow manual starts of each of the modified workflows, including an input to select a specific seed, perhaps to re-run a previous failed test.

burnpanck commented:
I forgot to mention; the `workflow_dispatch` is only effective if it is on the default branch, so we unfortunately won't be able to test that until after the merge.

burnpanck commented:
Oh, and the failed `GCC-13 C++23 fmtlib none Release` CMake test doesn't appear to be related to my changes here; either it has been pulled in with the recent merge, or it has always been there and has been discovered here through a lucky random choice.

mpusz commented:
> I also added a [workflow_dispatch](https://docs.github.com/en/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows#workflow_dispatch) trigger to allow

Thanks a lot! You are a true GitHub Actions wizard 🧙🏻 I wouldn't know how to do at least half of the things you contributed in the last few days.

mpusz commented on .github/generate-job-matrix.py:
```diff
@@ -18,25 +98,31 @@ def make_gcc_config(version: int) -> Configuration:
+        feature_support=Features(
+            cxx_modules=version >= 17,
+            std_format=version >= 17,
+            import_std=version >= 17,
```
According to our [compatiblity table](https://mpusz.github.io/mp-units/latest/getting_started/cpp_compiler_support/):

```suggestion
            import_std=version >= 18,
```

mpusz commented:
As you have introduced "features" maybe we should also explicitly name [`no_crtp`](https://mpusz.github.io/mp-units/latest/getting_started/cpp_compiler_support/#explicit-this-parameter)?

mpusz commented:
Thinking a bit more about `workflow_dispatch`, do you think it could be a good idea to be able to manually run the tests for all of the valid possible combinations? Being able to run it by hand could allow us to ensure that everything is OK before doing the library release.

mpusz commented:
@burnpanck, you have 5 open PRs. Do you plan to work on those soon? I would like to provide the next mp-units release soon.

burnpanck commented:
@mpusz: Unfortunately, I can't guarantee anything right now. I think most of my PRs were generally mergeable, but of course, now that time has passed, neither you nor I myself remember exactly what the last status was. This one here was ready except for your latest change requests, which do not appear too difficult to do. Maybe this weekend?

mpusz reviewed (dismissed):
Thanks! Please let me know if you plan to work a bit more on this or if it should be merged?

burnpanck commented:
Picking things up again... It looks like we got some new conan options that weren't implemented as features previously (`no_crtp` was already mentioned in this thread, then we have `natural_units`). The version of the GHA workflow in this branch compared to current master already has a few improvements to better separate "features" (conan options) from other build configurations; this should help both for these two features and future changes. I'll try to finish this and also implement that workflow dispatch with the option to run a denser or even full matrix.
